### DISCUSSION TITLE
Ruby on Rails(3.0)(6.1.7)で.add_row do |row|の部分でnil can't be coerced into Floatエラーとなる

### DISCUSSION THREAD

**Discussion by kotoyukihana** [thinreports/thinreports#36] [Q&A] [answered]
説明不足でしたすみません。
下記のレポート生成コード、.add_row do |row|の部分でnil can't be coerced into Floatエラーになります。

- ac.member.company.name
- ac.member.name
- ac.member.name
- index+1
↑がnilでないことは確認できています。

どこを確認すれば良いか、どのような確認が必要か、少しでも手掛かりが欲しいです。

class TestReport < BaseReport
  def initialize(test_class,preview_param)
    @test_class = test_class
    @preview_param=preview_param
    @user_class=UserClass.all
  end

  private

  attr_reader :test

  def template_name
    'test'
  end

  def build

    report.start_new_page
    page = report.page

    @user_class.each_with_index do |ac , index|
        page.list(:name_list).add_row do |row|
        row.item(:company).value(ac.member.company.name)
        row.item(:name).value(ac.member.name)
        row.item(:number).value(index+1)
      end
    end
  end
end

**Comment by kotoyukihana** [comment]
デバックの結果page.list(:name_list).add_row do |row|のブロック内ではなく、page.list(:name_list).add_row do |row|自体に問題がありそうです。.tlfファイルのバージョンは0.12.0で、page.list(:name_list)の内部はそれが反映されています。ただし、かなりのコード量があるのでどこが問題を抱えているのかわかりません。
どのような原因で、nil can't be coerced into Floatになるのでしょうか。

**Comment by kotoyukihana** [comment]
.add_rowはリスト内の項目を演算（数えて）して行を追加するような処理なのでしょうか。
行を追加するという制御でnil can't be coerced into Floatが発生する原因がわかりません。

**Comment by kotoyukihana** [comment]
解決しました。
thinrepots-Editerで出力した.tlfファイル内、"page-footer"、及び "footer"の"height":　がありませんでした。
（Editer上には表示されている）

そのため、page-footer heightがnilになり、下記gem内部            
　　　　def page_max_height
              @page_max_height ||= begin
                h = format.height
→             h -= format.section_height(:page_footer)
                h -= format.section_height(:footer) unless auto_page_break?
                h
              end
でnil can't be coerced into Floatエラーになりました。

以下は.tlfファイル内部
      "page-footer": {
        "enabled": true,
        "translate": {
          "x": 0,
          "y": 0
        },
        "items": []
      },
      "footer": {
        "enabled": true,
        "translate": {
          "x": 0,
          "y": 0
        },
        "items": []
      },
      "content-height": 751.3,
      "auto-page-break": true
      
      バージョン0.7.6の.tlfファイルを変換し、0.12.0にした結果かと思いますが、上記対応で当該エラーは解決しました。
      その他レポートも同様の修正で試みます。

**Comment by altivi** [comment, marked as answer]
@kotoyukihana 
提供いただいたコードにおいて、エラーメッセージ「nil can't be coerced into Float（nilはFloatに変換できません）」が.add_rowのブロック内ではなく、.add_row自体に関連しているようです。.tlfファイルのバージョンは0.12.0であり、page.list(:name_list)の内部もそれに対応しているようです。しかし、コードベースが非常に大きいため、具体的な問題を特定することは難しいです。

ご報告いただいた結果、.tlfファイルの「page-footer」と「footer」の部分に「height」プロパティが存在せず、それがエディタ上では表示されているにも関わらず、問題の原因となっているようです。そのため、gem内のpage_max_heightメソッドがnilの値に遭遇する可能性があります。

この問題を解決するためには、.tlfファイルに「page-footer」と「footer」の「height」プロパティを追加する必要があります。例えば以下のように修正します：

```
"page-footer": {
  "enabled": true,
  "translate": {
    "x": 0,
    "y": 0
  },
  "items": [],
  "height": 50  // ここにheightプロパティを追加
},
"footer": {
  "enabled": true,
  "translate": {
    "x": 0,
    "y": 0
  },
  "items": [],
  "height": 30  // ここにheightプロパティを追加
},
```
これらの修正を行うことで、「nil can't be coerced into Float」エラーが解消されるはずです。

ただし、提供された情報に基づいて、「height」プロパティの不足がエラーの原因であるという前提で指示をしています。もし問題が解決しない場合や、追加の詳細やエラーメッセージがある場合は、それらを提供していただければ、さらに具体的なサポートを提供できます。

役に立った場合は、これを回答としてマークしてください。
